webapp/platform/client/src/websocket.ts

```diff
@@ -135,6 +140,55 @@ export default class WebSocketClient {
             this.postedAck = postedAck;
         }
 
+        // Setup network event listener
+        if (typeof window !== 'undefined') {
+            // Remove existing listeners if any
+            if (this.onlineHandler) {
+                window.removeEventListener('online', this.onlineHandler);
+            }
+            if (this.offlineHandler) {
+                window.removeEventListener('offline', this.offlineHandler);
+            }
+
+            this.onlineHandler = () => {
+                // If we're already connected, don't need to do anything
+                if (this.conn && this.conn.readyState === WebSocket.OPEN) {
+                    return;
+                }
+
+                console.log('network online event received, scheduling reconnect'); // eslint-disable-line no-console
+
+                // Set a timer to reconnect after a delay to avoid rapid connection attempts
+                this.clearReconnectTimeout();
+                this.reconnectTimeout = setTimeout(
+                    () => {
+                        this.reconnectTimeout = null;
+                        this.initialize(connectionUrl, token, this.postedAck);
+                    },
+                    this.config.minWebSocketRetryTime,
+                );
+            };
+
+            this.offlineHandler = () => {
+                // If we've detected a full disconnection, don't need to do anything more
+                if (this.conn && this.conn.readyState !== WebSocket.OPEN) {
+                    return;
+                }
+
+                console.log('network offline event received, checking connection'); // eslint-disable-line no-console
+
+                // If we haven't detected a full disconnection,
+                // send a ping immediately to test the socket
+                this.waitingForPong = true;
+                this.ping(() => {
+                    this.waitingForPong = false;
+                });
```

Comment:
Can this interfere with any other ping already "in the air" (or the ping interval)?

I know there is a low chance for this, but let's imagine:
- The server/network is relatively slow, so the ping doesn't come back until 500ms have passed
- At 29s750ms since the last ping interval execution, we "lose internet".
- We set waiting for pong to be true, and send the ping
- The ping is not expected to arrive until 30s250ms, so at the 30s mark, the ping interval executes
- The ping interval sees that we are still waiting for the ping, so it closes the connection

Maybe the "risk" of that happening is only that we restart the connection when there was no need, so this may be fine like this, but I wonder if you had that scenario in mind.

Reply:
This is true, and I didn't think about it. Nice catch :+1:

I'm not sure it's worth worrying about, though. At a response of 500ms, we have a 1 in 60 chance that a random network "offline" event would cause a disconnect before we would expect the PONG. At a response of 100ms, that drops down to 1 in 300.

In addition to that, in most cases when we get an "offline" event we really are offline. So this would only affect the rare case that we somehow get an "offline" event without really being offline _and_ we hit the interval at the worst time. And in this case, we'd just try to reconnect.

WDYT?

Reply:
I agree that most of the time this won't be an issue. Let's just add a comment for the posterity so our future ourselves know that we thought about it, and why we decide to ignore it :P